### ISSUE TITLE
build docker image

### ISSUE THREAD

**Issue by willkg** [mozilla/socorro-pigeon#7]
We want a docker image that we can deploy our lambda code into and then run tests on.

**Comment by willkg** [comment]
I can't tell how deep in the woods I'm getting.

@milescrabill suggested https://github.com/lambci/docker-lambda which has a Python 2.7 set of images. 

I was thinking maybe we start with that image, install Ansible and then run the Ansible deploy playbook against localhost creating a new image we'll call pigeon-lambda or something like that.

Then we launch a rabbitmq container and the pigeon-lambda container running the tests and verifying the results.

Does that sound crazy? Are there better ways to do this such that we test the thing we're deploying effectively and can run this in CI?

**Comment by milescrabill** [comment]
It's a bit deep in the woods. Instead of testing the deploy, I think the Docker image should just have pigeon ready on it. There's a lot of AWS specific stuff happening in the Ansible playbook that we can deal with separately.

Testing against a rabbitmq container in the way you described makes sense. Testing in CircleCI with docker-compose should work.

**Comment by willkg** [comment]
To clarify, when I said "deploy", I meant all the steps except the last two which are amazon-specific. Seems like a good idea to package it into a `.zip` and then test against the contents of the `.zip`.

Relatedly, I've got most of this working now. Working on the test runner now.

**Comment by willkg** [comment]
Our deploy system uses Ansible, fetches requirements, generates a `pigeon.zip` file and then uses `lambda_function` to set everything up. That doesn't use `requirements.txt` or pip to install dependencies.

Thus to test pigeon correctly (where "correctly" is defined here as "we run the tests and have a lot of confidence that when we deploy this, it's the thing we tested and therefore it'll work"), we need to "deploy" pigeon into some AWS-lambda-like environment which picks up pigeon and the requirements and then run the tests on that deployed pigeon.

I spent a lot of time with `lambci/docker-lambda`. It's a family of docker containers for running AWS lambda things. There is a complete node.js environment that does what I would want, but we don't use node.js, so that's not helpful. The Python environment is nascent and missing stuff and doesn't work. I've been working through what's there to figure out what's missing. The runtime mock is run by `/var/runtime/awslambda/bootstrap.py` which does some setup I don't understand to run the lambda function.

At this point, I took a step back to reevaluate the options.

I think there are a couple:

1. Spend some more quality time to fix the `lambci/docker-lambda` Python runtime to run things. I'm new to AWS Lambda and don't understand what's missing so I'm not sure how long that will take. A week? We could use the fruits of this labor to keep moving forward, plus we could make a PR and upstream it.
2. Switch to another system. None of the other systems will test our deploy or make sure we've got requirements set up right. That's probably fine if we use -dev to catch these errors which I think this aligns with what @milescrabill is thinking. Couple of suboptions here:
  1. There are a few other "run something in an AWS lambda-esque environment" things out there. We could switch to one of those. I did a quick survey of them, but wasn't immediately fond of any. Some require AWS setup. Some have APIs that are not convenient for tests. Some are "complete dev, test and deploy" systems which I think would get in the way of what we do for dev and deploy.
  2. I could roll my own system based on one of the ones I found. Maybe a py.test fixture for testing convenience?

In writing this, I think I'm going to find another system that's close to what I want and then if I haven problems, I'll know enough to write a py.test fixture that's good enough.
